### ISSUE TITLE
Compilation error for code with forward reference hint 

### ISSUE THREAD

**Issue by petrstupka** [cython/cython#1887]
Hello,
when trying to compile following code (simplified example):
```
class A:
    def __init__(self, x):
        self.x = x
    def get_x(self) -> 'X':
        return self.x

class X:
    pass
```
(with cython 0.27) I get error message
```Error compiling Cython file:
------------------------------------------------------------
...
    def __init__(self, x):
        self.x = x
    def get_x(self) -> 'X':
        return self.x

class X:
^
------------------------------------------------------------

test.py:7:0: Assignment to non-lvalue 'X'
```
After removing forward reference type hint `-> 'X'` from `A.get_x()` method, compilation proceeds without error.  
It seems more like a parser's problem.

**Comment by scoder** [comment]
Thanks for the report. What happens is that Cython looks up `X` as a type in the annotation, does not find it (yet), and assumes that it's something from an outer scope, i.e. a builtin. When it then finds the class definition, it tries to assign the class object to the assumed builtin `X`, and rejects it as invalid.

**Comment by scoder** [comment]
Fixed in [0.27.1b1](https://github.com/cython/cython/archive/0.27.1b1.tar.gz). Please give it a try.

**Comment by petrstupka** [comment]
Hi,
thank you very much for the fix. I have already checked 0.27 version with your patch applied. It works for test case and also works for all real cases occuring in my python projects.
